server/src/grpc/remote_engine_service/mod.rs

```diff
@@ -97,6 +155,139 @@ impl<Q: QueryExecutor + 'static> RemoteEngineServiceImpl<Q> {
+            RequestResult::Wait => {
+                REMOTE_ENGINE_GRPC_HANDLER_DURATION_HISTOGRAM_VEC
+                    .stream_read
+                    .observe(instant.saturating_elapsed().as_secs_f64());
+                return Ok(ReceiverStream::new(rx));
+            }
+        };
+
+        let handle = self
+            .runtimes
+            .read_runtime
+            .spawn(async move { handle_stream_read(ctx, request).await });
+        let streams = handle.await.box_err().context(ErrWithCause {
+            code: StatusCode::Internal,
+            msg: "fail to join task",
+        })??;
+
+        let mut stream_read = FuturesUnordered::new();
+        for stream in streams.streams {
+            let mut stream = stream.map(|result| {
+                result.box_err().context(ErrWithCause {
+                    code: StatusCode::Internal,
+                    msg: "record batch failed",
+                })
+            });
+
+            let handle = self.runtimes.read_runtime.spawn(async move {
+                let mut batches = Vec::new();
+                while let Some(batch) = stream.next().await {
+                    batches.push(batch)
+                }
+
+                batches
+            });
+            stream_read.push(handle);
+        }
+
+        let mut batches = Vec::new();
+        while let Some(result) = stream_read.next().await {
+            let batch = result.box_err().context(ErrWithCause {
+                code: StatusCode::Internal,
+                msg: "failed to join task",
+            })?;
+            batches.extend(batch);
+        }
+
+        // We should set cancel to guard, otherwise the key will be removed twice.
+        guard.cancel();
+        let notifiers = self
+            .request_notifiers
+            .as_ref()
+            .unwrap()
+            .take_notifiers(&request_key)
+            .unwrap();
+
+        let mut num_rows = 0;
+        for batch in batches {
+            match batch {
+                Ok(batch) => {
+                    num_rows += batch.num_rows() * notifiers.len();
+                    for notifier in &notifiers {
+                        if let Err(e) = notifier.send(Ok(batch.clone())).await {
+                            error!("Failed to send handler result, err:{}.", e);
+                        }
+                    }
+                }
+                Err(_) => {
+                    for notifier in &notifiers {
+                        let err = ErrNoCause {
+                            code: StatusCode::Internal,
+                            msg: "failed to handler request".to_string(),
+                        }
+                        .fail();
+                        if let Err(e) = notifier.send(err).await {
+                            error!("Failed to send handler result, err:{}.", e);
+                        }
+                    }
+                    break;
+                }
+            }
+        }
+
```

Comment:
After the succeeds in notifying, how about adding a metric to record the number of `dedupped_stream_query`?